src/test/java/com/google/cloud/spanner/pgadapter/ITErrorsTest.java

```diff
@@ -0,0 +1,172 @@
+                "create unique index my_index on my_table (value)",
+                "create table my_other_table (id bigint primary key, my_table_id bigint, constraint fk_my_table foreign key (my_table_id) references my_table (id))"));
+    testEnv.startPGAdapterServer(Collections.emptyList());
+  }
+
+  @AfterClass
+  public static void teardown() {
+    testEnv.stopPGAdapterServer();
+    testEnv.cleanUp();
+  }
+
+  @After
+  public void clearTestData() throws SQLException {
+    try (Connection connection = DriverManager.getConnection(getConnectionUrl())) {
+      connection.createStatement().execute("delete from my_other_table");
+      connection.createStatement().execute("delete from my_table");
+    }
+  }
+
+  private String getConnectionUrl() {
+    return String.format(
+        "jdbc:postgresql://%s/%s",
+        testEnv.getPGAdapterHostAndPort(), database.getId().getDatabase());
+  }
+
+  @Test
+  public void testTableNotFound() throws SQLException {
+    try (Connection connection = DriverManager.getConnection(getConnectionUrl())) {
+      PSQLException exception =
+          assertThrows(
+              PSQLException.class,
+              () ->
+                  connection
+                      .createStatement()
+                      .executeUpdate("insert into my_non_existing_table (id, ref) values (1, 1)"));
+      assertNotNull(exception.getServerErrorMessage());
+      assertEquals(
+          SQLState.UndefinedTable.toString(), exception.getServerErrorMessage().getSQLState());
+      assertEquals(SQLState.UndefinedTable.toString(), exception.getSQLState());
+    }
+  }
+
+  @Test
+  public void testColumnNotFound() throws SQLException {
+    try (Connection connection = DriverManager.getConnection(getConnectionUrl())) {
+      PSQLException exception =
+          assertThrows(
+              PSQLException.class,
+              () ->
+                  connection
+                      .createStatement()
+                      .executeUpdate("insert into my_other_table (id, ref) values (1, 1)"));
+      assertNotNull(exception.getServerErrorMessage());
+      assertEquals(
+          SQLState.UndefinedColumn.toString(), exception.getServerErrorMessage().getSQLState());
+      assertEquals(SQLState.UndefinedColumn.toString(), exception.getSQLState());
+    }
+  }
+
+  @Test
+  public void testPrimaryKeyConstraintViolation() throws SQLException {
+    try (Connection connection = DriverManager.getConnection(getConnectionUrl())) {
+      assertEquals(
+          1,
+          connection
+              .createStatement()
+              .executeUpdate("insert into my_table (id, value) values (1, 'One')"));
+      PSQLException exception =
+          assertThrows(
+              PSQLException.class,
+              () ->
+                  connection
+                      .createStatement()
+                      .executeUpdate("insert into my_table (id, value) values (1, 'Other one')"));
```

Comment:
Can we create a private method and pass 2 things from each test - sql query and the SQLstate enum as thats the only difference in all tests?

Reply:
Added a separate private method to verify the expected SQL state.

We use Java 8 with JUnit4, so support for parameterized tests is very limited (it supports parameterized test classes, but not parameterized tests).